utils_cv/detection/model.py

```diff
@@ -247,18 +380,105 @@ def plot_precision_loss_curves(
         ax2.set_ylabel("average precision", color="b")
         ax2.plot(self.ap, "b-")
 
-    def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
-        """ eval code on validation/test set and saves the evaluation results in self.results. """
+    def evaluate(self, dl: DataLoader = None) -> Union[CocoEvaluator, None]:
+        """ eval code on validation/test set and saves the evaluation results
+        in self.results.
+        """
         if dl is None:
+            if not self.dataset:
+                return
             dl = self.dataset.test_dl
         self.results = evaluate(self.model, dl, device=self.device)
         return self.results
 
+    def _transform(
+        self,
+        im: Union[str, Path, Image.Image, np.ndarray],
+    ) -> torch.Tensor:
+        """ Convert the image to the format required by the model. """
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im)
+        if self.device:
+            im.to(self.device)
+        return im
+
+    @classmethod
+    def _apply_threshold(
+        cls,
+        pred: Dict,
+        threshold: Optional[int] = 0.5,
+        mask_threshold: Optional[int] = 0.5,
+        **kwargs,
+    ) -> Dict:
+        """ Return prediction results that are above the threshold if any. """
+        # detach prediction results to cpu
+        pred = {k: v.detach().cpu().numpy() for k, v in pred.items()}
+        # apply score threshold
+        if threshold:
+            selected = pred['scores'] > threshold
+            pred = {k: v[selected] for k, v in pred.items()}
+        if "masks" in pred and mask_threshold:
+            pred["masks"] = pred["masks"] > mask_threshold
+        return pred
+
+    def _get_det_bboxes(
+        self,
+        pred: Dict,
+        im_path: Union[str, Path]
+    ) -> List[Type[_Bbox]]:
+        return DetectionBbox.from_arrays(
+            pred['boxes'].tolist(),
+            score=pred['scores'].tolist(),
+            label_idx=pred['labels'].tolist(),
+            label_name=np.array(self.labels)[pred['labels']].tolist(),
+            im_path=im_path,
+        )
+
+    def _process_pred_results(
+        self,
+        pred: Dict,
+        im_path: Union[str, Path]
+    ) -> Union[List[Type[_Bbox]], Tuple]:
+
+        res = self._get_det_bboxes(pred, im_path)
+        if "masks" in pred:
+            res = (res, pred["masks"].squeeze())
+        elif "keypoints" in pred:
+            res = (res, pred["keypoints"])
+        return res
+
+    @classmethod
+    def _pack_pred_results(cls, res: List, infos: Dict) -> List[Dict]:
```

Comment:
can you add docstring stating what this function does?

Reply:
Done